### PR TITLE
feat: introspection context

### LINKED ISSUE #39: Introspection Context
Issue to implement an Introspection Context, as described in https://github.com/strimzi/strimzi-ui/blob/master/docs/Architecture.md#introspection and https://github.com/strimzi/strimzi-ui/tree/master/client/Contexts/Introspect

- Requires GQL library ( #38 ), testing issues to be completed

### PR DESCRIPTION
**Pull request by pmuir**
Fixes #39

Signed-off-by: Pete Muir <pmuir@bleepbleep.org.uk>

### CODE REVIEW

**Comment by pmuir** [comment]
Adding this to get some early comments about whether this is the right direction or not.

This introspection handles the example in the readme. I need to add support for non Scalar types to the field checked. We possibly also want to consider passing some arguments to callbacks to allow them to evaluate whether the operation is supported. We may also want to check if particular types are attached to the query and mutation.

**Comment by pmuir** [comment]
I had to import the graphql dependency but we can fix that up after we merge the GraphQL PR.

**Review by matthew-chirgwin — COMMENTED** [review]
Thanks for this - looks exactly like what I was expecting. A few comments so far:

- Rather than implement .js files, could they be .ts (and define types/interfaces) for others to consume as appropriate?
- I assume you planning to add the React context producer/consumer pieces in a following commit?
  - As well as tests?
> We possibly also want to consider passing some arguments to callbacks to allow them to evaluate whether the operation is supported. 
- What scenarios do you see which may need this (in the scope of introspection)? I absolutley agree that the result of the introspection, security (#44) etc will need to be combined (as described in the page here: https://github.com/strimzi/strimzi-ui/tree/master/client/Pages - backendSupportFor in the case of introspection )

**Comment by pmuir** [comment]
> Rather than implement .js files, could they be .ts (and define types/interfaces) for others to consume as appropriate?

Sure, I just used the js as existing files in `client` were js.

> I assume you planning to add the React context producer/consumer pieces in a following commit?

Right - I wanted to get feedback on actual meat of the code first.

I just pushed better field validation code.

> As well as tests?

Yes, I will write a simple harness that can read in sample introspection results and expectations, and process them. This will allow us to cover many different schemas/expectations. I'll start with the schema/expectations I've been validating with. 

I'll do this next.

> We possibly also want to consider passing some arguments to callbacks to allow them to evaluate whether the operation is supported.
> 
> What scenarios do you see which may need this (in the scope of introspection)? I absolutley agree that the result of the introspection, security (#44) etc will need to be combined (as described in the page here: https://github.com/strimzi/strimzi-ui/tree/master/client/Pages - backendSupportFor in the case of introspection )

I added some examples of this to the samples, but if we defined a `topicsByName` query, then we might want to check it's arguments to see if it supports certain arguments e.g.

```findByName: ({ queries }) => {
        return (
          queries['topicsByName'] !== undefined &&
          queries['topicsByName'].args.reduce((acc, curr) => {
            if (curr.name === 'name') {
              return true;
            } else {
              return acc;
            }
          }, false)
        );
      },```

File: client/Contexts/index.js
Comment (nictownsend) [inline client/Contexts/index.js]: `.ts` instead?

**Comment by pmuir** [comment]
This is ready for review - I've had to base it on top of @JQrdan 's GraphQL PR and that should merge first.

**Comment by matthew-chirgwin** [comment]
@pmuir the client has now been merged - once rebased I will review this

File: client/Contexts/Introspect/Expectations.ts
Comment (nictownsend) [inline client/Contexts/Introspect/Expectations.ts]: Could possibly avoid duplication with a partial function:

```
const mutationValidator = (operationName) => ({mutations}) => mutations[operationName] !== undefined;
...
create : mutationValidator('createTopic')
```

File: client/Contexts/Introspect/Expectations.ts
Comment (nictownsend) [inline client/Contexts/Introspect/Expectations.ts]: `return` will just exit the loop but not shortcut the reduce so a later false would result in accumulator having false.

I think you need:
```
return (curr || acc) ? true : false;
```

 You also need to seed the reduce with a default value in case there are no args - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Reduce_of_empty_array_with_no_initial_value

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: Array of `T`, reduce into a single object hash whereby each `T` is stored using its value of `by` as the key in the hash - what if value of `by` is the same/missing for multiple items? And does TS handle cases where the value isn't a string?

Also I think referring to `index` is confusing - it's not the index of the array, it's the prop name to use as the key.

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: This seems to be indexing a single "thing" from the schema, not the whole schema. Misleading function name?

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: What if the type can't be found? NPE as type is `undefined`

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: NPE if type undefined

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: Do you need to define the shape with defaults? For e.g `answer.fields` may be left undefined below.

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (nictownsend) [inline client/Contexts/Introspect/mock/expectations-basic.ts]: Minor concerns that this duplicates a lot of the real `Expectations` - is it needed or can it be reduced to provide a smaller test case (e.g the reducers in the callbacks)

**Review by nictownsend — CHANGES_REQUESTED** [review]

File: client/Contexts/Introspect/Expectations.ts
Comment (pmuir) [reply client/Contexts/Introspect/Expectations.ts]: To be clear, I don't expect this to be the *real* expectations for `Topic`, so it's probably better to actually just remove the contents.

File: client/Contexts/Introspect/introspection.ts
Comment (pmuir) [reply client/Contexts/Introspect/introspection.ts]: > Array of T, reduce into a single object hash whereby each T is stored using its value of by as the key in the hash - what if value of by is the same/missing for multiple items?

I don't think it can be, in the usage of this function here (as the introspection schema won't return the same thing twice), but I'll add an explicit check and document it.

> And does TS handle cases where the value isn't a string?

I don't think it does so I've added an explicit check and documented it. Again, I don't believe in the usage of the function this error can occur (due to what the introspection schema returns).

> Also I think referring to index is confusing - it's not the index of the array, it's the prop name to use as the key.

Used your text.

File: client/Contexts/Introspect/introspection.ts
Comment (pmuir) [reply client/Contexts/Introspect/introspection.ts]: Yeah, I really struggled with what to call this function. Suggestions?

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (pmuir) [reply client/Contexts/Introspect/mock/expectations-basic.ts]: Yeah, I removed the real Expectations - we haven't actually determined the real ones yet. This one purely exists for unit testing this code. We don't want to merge them as that would mean that changes to the real expectations incorrectly break tests.

File: client/Contexts/Introspect/introspection.ts
Comment (pmuir) [reply client/Contexts/Introspect/introspection.ts]: Yes, I think that makes sense.

File: client/Contexts/Introspect/Expectations.ts
Comment (pmuir) [reply client/Contexts/Introspect/Expectations.ts]: How would it return false later?

Comment (pmuir) [reply client/Contexts/Introspect/Expectations.ts]: Anyway, I changed it to your form as that also works.

File: client/Contexts/Introspect/Expectations.ts
Comment (pmuir) [reply client/Contexts/Introspect/Expectations.ts]: I removed it.

**Review by dlabaj — APPROVED** [review]

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (nictownsend) [inline client/Contexts/Introspect/mock/expectations-basic.ts]: This comment depends on whether I've interpreted the intention of the code.

If this interpretation is correct:

```
- Iterate over the list of args, mapping to `true|false` based off whether the arg is called `"name"`
- Reduce that mapped array into a single boolean, which will be `true` if any of the args were called `name`
```

If that is the case - I think the original code had a `return false` if the current value was false. That meant that if the first arg mapped to `true`, but the remaining `n` mapped to `false` - the reduce result would be `false` (I don't think you can short circuit a `reduce` so the first `true` would later be overridden by the `false`s).

In addition; if my description is correct, you can simplify it with the following:

```
queries['topicsByName'].args.findIndex(v => v.name === 'name') > -1
```

Comment (nictownsend) [reply client/Contexts/Introspect/mock/expectations-basic.ts]: I also note that this is now only in the mocks (but probably still valid in case this becomes a seed for the real expectations when later added back)

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]:
```suggestion
const indexTypeFromSchema = (
```
maybe? naming is hard :)

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: https://lodash.com/docs/4.17.15#keyBy - might simplify code with one caveat - it will override duplicate keys with latest value, rather than throw error (depending on whether that's important).

```suggestion
  return _keyBy(array, thing => {
  const prop = thing[propName];
    if (typeof prop !== 'string') {
      throw new Error(
        `key identified by ${propName} in ${curr} must be of type "string"`
      );
    }
    return prop;
```

**Review by nictownsend — CHANGES_REQUESTED** [review]
These are pretty much semantic changes rather than blockers - so if you think it's not worth changing I'll change to approval.

File: client/Contexts/Introspect/introspection.ts
Comment (nictownsend) [inline client/Contexts/Introspect/introspection.ts]: This will need an install of lodash's keyBy function package

**Review by nictownsend — CHANGES_REQUESTED** [review]

**Review by nictownsend — APPROVED** [review]

File: package.json
Comment (matthew-chirgwin) [inline package.json]: Looks like a rebase change - we were at 1.29, but after this would be a 1.26.10

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (matthew-chirgwin) [inline client/Contexts/Introspect/mock/expectations-basic.ts:21]: Is this just to illustrate a type can reference others? If so, would doc be a suitable place for it? Might save on the TODO too?

**Review by matthew-chirgwin — CHANGES_REQUESTED** [review]
LGTM! Spotted a small (rebase) dependency change, and had one question around the `other` expectation defined

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (pmuir) [reply client/Contexts/Introspect/mock/expectations-basic.ts:21]: No, it's mock data used to test the introspector. It's important to validate that the instrospector works with all different types in GraphQL IMO.

File: package.json
Comment (pmuir) [reply package.json]: this is caused by mistake in the current `package.json` - see https://github.com/strimzi/strimzi-ui/blob/master/package.json#L66-L68 - `sass` is specified twice.

Comment (pmuir) [reply package.json]: I can fix it to use 1.29 though.

File: client/Contexts/Introspect/mock/expectations-basic.ts
Comment (pmuir) [reply client/Contexts/Introspect/mock/expectations-basic.ts:21]: no, it's mock data used to validate the introspector

Comment (pmuir) [reply client/Contexts/Introspect/mock/expectations-basic.ts:21]: The TODO is valid, but we can just make an issue for it - I don't think it's a common case.

File: client/Contexts/Introspect/Context.tsx
Comment (jeff-phillips-18) [inline client/Contexts/Introspect/Context.tsx]: Prefer to just import `React` and prefix all the others rather than mixing.

File: client/Contexts/Introspect/Context.tsx
Comment (jeff-phillips-18) [inline client/Contexts/Introspect/Context.tsx]:
```suggestion
    const [result, setResult] = useState<IntrospectionResult>({
```

File: client/Contexts/Introspect/Context.tsx
Comment (jeff-phillips-18) [inline client/Contexts/Introspect/Context.tsx]:
```suggestion
    response: ApolloQueryResult<IntrospectionQuery> | null
```

File: client/Contexts/Introspect/Context.tsx
Comment (jeff-phillips-18) [inline client/Contexts/Introspect/Context.tsx]:
```suggestion
                response: null
            })
```

File: client/Contexts/Introspect/Context.tsx
Comment (jeff-phillips-18) [inline client/Contexts/Introspect/Context.tsx]:
```suggestion
            response: null
        })
```

**Review by jeff-phillips-18 — APPROVED** [review]
Some suggestions, otherwise lgtm

**Comment by pmuir** [comment]
I've pulled on the `Context` impl for now, will add it once #99 merges with the test utility for graphql fetching.

**Review by matthew-chirgwin — APPROVED** [review]
Changes LGTM - thanks :)

Wont block the PR on it, but could a brief summary of the files in the `mock` directory be added to the Introspect README in a following PR? Just to cover what they are, how to go about updating them (if required) etc?

**Review by scholzj — APPROVED** [review]

**Pull request merged**
